File: 22-generate-parentheses/generate-parentheses.java
class Solution {
    public void g(int n,int l, int r, String s,List<String> k)
    {
        if(r == n)
        {
            k.add(s);
            return;
        }
        if(l < n)
        {
            g(n,l+1,r,s+"(",k);
        }
        if(r < l)
        {
            g(n,l,r+1,s+")",k);
        }
    }
    public List<String> generateParenthesis(int n) {
        List<String> l = new ArrayList<>();
        g(n,0,0,"",l);
        return l;
    }
}

File: 22-generate-parentheses/generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []
        self.ss(res,"",0,0,n)
        return res

    def ss(self, res, s, o, c, ma):
        if len(s) == ma*2:
            res.append(s)
            return
        
        if o < ma:
            s += "("
            self.ss(res,s,o+1,c,ma)
            s = s[:-1]
        
        if c < o:
            s += ")"
            self.ss(res,s,o,c+1,ma)
            s = s[:-1]

        